### PR TITLE
HPCC-14936 EclWatch logical file copy to other clusters, loses blockCompression flag and adds other spurious meta flags.

### PR DESCRIPTION
**Pull request by AttilaVamos**
Fix compressed file copy issues
Fix/restict header/footerlength attributes insertion.

Signed-off-by: Attila Vamos attila.vamos@gmail.com

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith please review.

File: dali/ft/filecopy.cpp [hunk 1]
```diff
@@ -1224,6 +1224,8 @@ void FileSprayer::calculateSprayPartition()
                 distributedTarget->setECL(recStru.str());
         }
     }
+    if (compressedInput && compressOutput && streq(encryptKey.str(),decryptKey.str()))
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: I assume encryptKey corresponds to the compressed output? and decryptKey, corresponds to the compressed input ?
Couldn't they differ?
Couldn't the output be compressed, but the input not?

File: dali/ft/filecopy.cpp [hunk 2]
```diff
@@ -2991,7 +2993,7 @@ void FileSprayer::updateTargetProperties()
             offset_t physPartLength = curProgress.outputLength;
             if (copyCompressed) {
                 FilePartInfo & curSource = sources.item(cur.whichInput);
-                partLength = curSource.size;
+                partLength = physPartLength;
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: physPartLength (curProgress.outputLength) is the uncompressed part size?

vs curProgress.compressedPartSize, which is the compressed part size.

If that's correct, I think that physPartLength variable should be renamed, as a little confusing at the moment. i.e. it is not the 'physical part length' if compressed.

File: dali/ft/filecopy.cpp [hunk 3]
```diff
@@ -3136,12 +3131,17 @@ void FileSprayer::updateTargetProperties()
                      ((stricmp(aname,FArecordCount)==0)&&!gotrc) ||
                      ((stricmp(aname,"@blockCompressed")==0)&&copyCompressed) ||
                      ((stricmp(aname,"@rowCompressed")==0)&&copyCompressed)||
-                     (stricmp(aname,"@local")==0)
+                     (stricmp(aname,"@local")==0)||
+                     (stricmp(aname,"@checkSum")==0)||
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: Is it correct to copy @checkSum?

could (does it need to) the calculated check sum as a result of spray ?

**Comment by jakesmith** [comment]
@AttilaVamos - a few questions

File: dali/ft/filecopy.cpp (hunk 1 above)
Comment (AttilaVamos) [inline dali/ft/filecopy.cpp]: Theoretically yes, they can be differ from DFUPlus help:
        encrypt=<password> -- optional, encrypt target
        decrypt=<password> -- optional, decrypt source

Yes, you can copy uncompressed source file to a compressed target.

File: dali/ft/filecopy.cpp (hunk 2 above)
Comment (AttilaVamos) [inline dali/ft/filecopy.cpp]: I will update this soon.

File: dali/ft/filecopy.cpp (hunk 3 above)
Comment (AttilaVamos) [inline dali/ft/filecopy.cpp]: Actually the spray/copy doesn't calculate CRC if the spray/copy target compressed or copy compressed source file to compressed target, based on  this function:

bool FileSprayer::calcCRC()
{
    return options->getPropBool(ANcrc, true) && !compressOutput && !copyCompressed; 
}

**Comment by AttilaVamos** [comment]
I found a bug(s) in compressed copy related original (uncompressed), compressed size stored into metainfo. They are wrong as well as the parts original and compressed size value.

**Comment by richardkchapman** [comment]
@jakesmith Can you re-review

File: dali/ft/filecopy.cpp (hunk 3 above)
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: Which makes me more dubious as to whether copying @checkSum unconditionally here is correct.

copying source @checkSum isn't going to be correct if file changed at all.

**Comment by jakesmith** [comment]
@AttilaVamos - 1 question re. @checkSum

**Comment by AttilaVamos** [comment]
@jakesmith updated.

**Comment by jakesmith** [comment]
@richardkchapman - looks good to merge.

**Pull request merged**